### ISSUE TITLE
Listen for updates to the unread marker

### ISSUE THREAD

**Issue by trygveaa** [weechat/weechat#2027]
## Feature description

As far as I can see, it's not currently possible to listen to changes to the unread marker. E.g. wee-slack needs this to update the server with the read state when it changes in WeeChat.

wee-slack currently does this by listening to the `buffer_switched` signal, and hooking the `/buffer set unread`/`/input set_unread*` commands, however this has several problems:

1. If the unread marker is changed from another plugin/script with the `buffer_set` function, it's not detected.
2. Hooking a command means that it runs before the actual command is run, so `last_read_line` isn't updated yet. This can be worked around by sleeping with `hook_timer`, but it's a bit cumbersome.
3.  If a new command to update the read marker is added (like it was in 3.8), the new command isn't detected until the script is updated.
4. You have to look up the read marker via hdata to see what it was changed to.
5. The `buffer_switched` signal doesn't tell you which buffer you switched from, so you have to keep track of that.
6. The read marker isn't necessarily changed when the buffer is switched, so you have to look up `first_line_not_read` and `last_read_line` to check if it was actually changed.

**Comment by trygveaa** [comment]
Merged buffers makes this even more complicated, as when `buffer_switched` or one of the commands are triggered, you have to check if there are any buffers merged with the buffer it triggered on, and check/update these as well.

**Comment by flashcode** [comment]
@trygveaa: so if I understand, your need would be to have a signal sent (with the buffer pointer as argument) whenever one of these variables changes in lines of a buffer:

- `first_line_not_read`
- `last_read_line`

Am I correct?

**Comment by trygveaa** [comment]
Yes, correct. It would be practical to get those values in the signal callback as well, so I don't have to look them up.

And I guess it might be a difference if `own_lines` or `mixed_lines` change, or do they always change at the same time?

**Comment by flashcode** [comment]
They should change at same time. In the code, only `lines->last_read_line` is used, `lines` being a pointer to `own_lines` for a single buffer and to `mixed_lines` for a buffer merged with others.

So the signal could be sent whenever `lines->last_read_line` or `lines->first_line_not_read` changes.

Sending these values with the signal is not so easy, as only a pointer, string or integer can be sent with a signal.
Instead a hsignal could be sent, which uses a hashtable as argument. Would it be OK?

Other ideas?

**Comment by trygveaa** [comment]
> They should change at same time. In the code, only `lines->last_read_line` is used, `lines` being a pointer to `own_lines` for a single buffer and to `mixed_lines` for a buffer merged with others.

Okay, thanks for the info.

> Sending these values with the signal is not so easy, as only a pointer, string or integer can be sent with a signal. Instead a hsignal could be sent, which uses a hashtable as argument. Would it be OK?
> 
> Other ideas?

Using a hsignal sounds good to me. Are there any downside versus using a signal?

**Comment by flashcode** [comment]
The only downside is that it's a bit slower to send and catch (build and copy of hashtable), but it's more flexible, in case we want to add fields in future.

**Comment by trygveaa** [comment]
Since this is something that typically happens on user interaction, and not something that will occur lots of times, I guess that performance difference is pretty negligible?
